tracing-subscriber/src/filter/layer_filters.rs

```diff
@@ -43,49 +83,117 @@ pub struct FilterFn<
     _s: PhantomData<fn(S)>,
 }
 
-#[derive(Copy, Clone, Debug)]
-pub struct FilterId(u8);
+/// A per-layer [`Filter`] implemented by a closure or function pointer that
+/// determines whether a given span or event is enabled, based on its
+/// [`Metadata`].
+///
+/// See the [documentation on per-layer filtering][plf] for details.
+///
+/// [`Metadata`]: tracing_core::Metadata
+/// [`Filter`]: crate::layer::Filter
+/// [plf]: crate::Layer#per-layer-filtering
+#[derive(Clone)]
+pub struct FilterFn<F = fn(&Metadata<'_>) -> bool> {
+    enabled: F,
+    max_level_hint: Option<LevelFilter>,
+}
+
+/// Uniquely identifies an individual [`Filter`] instance in the context of
+/// a [`Subscriber`].
+///
+/// When adding a [`Filtered`] [`Layer`] to a [`Subscriber`], the [`Subscriber`]
+/// generates a `FilterId` for that [`Filtered`] layer. The [`Filtered`] layer
+/// will then use the generated ID to query whether a particular span was
+/// previously enabled by that layer's [`Filter`].
+///
+/// **Note**: Currently, the [`Registry`] type provided by this crate is the
+/// **only** [`Subscriber`] implementation capable of participating in per-layer
+/// filtering. Therefore, the `FilterId` type cannot currently be constructed by
+/// code outside of `tracing-subscriber`. In the future, new APIs will be added to `tracing-subscriber` to
+/// allow non-Registry [`Subscriber`]s to also participate in per-layer
+/// filtering. When those APIs are added, subscribers will be responsible
+/// for generating and assigning `FilterId`s.
+///
+/// [`Filter`]: crate::layer::Filter
+/// [`Subscriber`]: tracing_core::Subscriber
+/// [`Layer`]: crate::layer::Layer
+/// [`Registry`]: crate::registry::Registry
+#[derive(Copy, Clone)]
+pub struct FilterId(u64);
```

Comment:
A filter ID _is_ a filter mask? Unless I'm missing something?

Reply:
I suppose my argument is that there is a difference between a `FilterId`, which is a unique identifier for each filter, and the AND of multiple `FilterId`s, which I would argue should have a dedicated type with dedicated documentation.

Reply:
I'll probably do this internal refactoring in a follow-up branch. I'd really like to get the code merged sooner rather than later. It's a good suggestion though!